Late temporal divergence lowering for SDAG

petar-avramovic opened the pull request:
D155343 introduced performance regressions by not allowing machine instruction sinking in some cases - SWEDEV-414443.
It originally fixed SWDEV-407790

Revert D155343 to fix performance regression SWEDEV-414443
and introduce temporal divergence lowering pass that also fixes SWDEV-407790

petar-avramovic committed:
Revert "MachineSink: Fix sinking VGPR def out of a divergent loop"

This reverts commit 3f8ef57bede94445b1a1042c987cc914a886e7ff.

petar-avramovic committed:
AMDGPU: Add test for temporal divergence introduced by machine-sink

Introduced by 5b657f50b8e8dc5836fb80e566ca7569fd04c26f that moved
LICM after AMDGPUCodeGenPrepare. Some instructions are no longer
sunk during ir optimizations but in machine-sinking instead.
If vgpr instruction used sgpr defined inside the loop is sunk outside
of the loop we end up with not-handled case of temporal divergence.

petar-avramovic committed:
AMDGPU: Fix temporal divergence introduced by machine-sink

Temporal divergence that was present in input or introduced in ir
transforms, like code-sinking or LICM is handled in SIFixSGPRCopies
by changing sgpr source instr to vgpr instr.
After 5b657f50b8e8dc5836fb80e566ca7569fd04c26f, that moved LICM after
AMDGPUCodeGenPrepare, machine-sinking can introduce temporal divergence
by sinking instructions out of loops.

Introduce new machine function pass that handles temporal divergence by
inserting copies to vgpr inside the loop.
Pass uses machine uniformity analysis to detect temporal divergent uses.

ssahasra commented on llvm/test/CodeGen/AMDGPU/machine-sink-temporal-divergence-swdev407790.ll:
Does the test have to be so big? It's really difficult to see which value is of interest in the LLVM IR. So if the FileCheck patterns are auto updated in the future, it will be difficult to determine if the tested thing is still correct. Also, in the future, it is possible that the test breaks for some completely different reason, that is not related to uniformity.

I would strongly recommended a much simpler hand-coded test which demonstrates what the actual pattern of interest.

ruiling replied:
We can also try to remove some unnecessary blocks but still show the problem. This might be easy to get an LLVM IR based lit-test. But the separate pass might need some hand-crafted MachineIR test to cover more cases.

ssahasra reviewed (commented):
I think the first line of the commit description should not mention machine sink. The issue was triggered by machine sink, but this change introduces a pass that works in general.

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUTemporalDivergenceLowering.cpp:
```diff
@@ -0,0 +1,121 @@
+  const SIRegisterInfo &TRI = *Subtarget.getRegisterInfo();
+
+  // Temporal divergence lowering is required for uniform UniformSourceReg
+  // and divergent UserInstr. UserInstr is uniform only when loop is uniform.
```
The changes should consistently use the term "cycle" and not "loop" everywhere.

What does this comment mean? Is it a pre-condition or a post-condition for this pass? By "loop is uniform", do you mean the cycle does not have divergent exits? Since UserInstr is outside the cycle, it could be divergent for any other reason too, like some operand which is not inside the cycle.

petar-avramovic replied:
It was meant to be a simple comment explaining when we need to insert a copy to vgpr, also it is there to give better context for fix me comment below. UserInstr in reported as uniform in some cases (cycle had divergent exit) 

I would expect that it was enough to ask if SrcReg was uniform (UserInstr should be divergent because of temporal divergence unless it was a uniform loop)

However, machine uniformity analysis detects all temporal divergence cases that require lowering. 

> By "loop is uniform", do you mean the cycle does not have divergent exits?
I am not sure about the terminology can you point me some reference? But at this point cycles have exactly one entry point (is that good enough to call them natural loops?) and exactly one exit. I only looked in tests where something was wrong and all of them had "divergent exit from the cycle"

ssahasra replied:
Ok, I finally parsed this statement correctly: "UserInstr is uniform only when loop is uniform." I got confused by the "only when". About the term "loop", how do you know that every cycle is reducible at this point? This pass is fairly generic, and could be moved around. If this pass actually assumes that every cycle is reducible, then there should be asserts about that. Also separately, instead of saying "is uniform", it's less confusing to say "has divergent exits".

ssahasra commented on llvm/include/llvm/ADT/GenericUniformityImpl.h:
```diff
@@ -396,6 +399,8 @@ template <typename ContextT> class GenericUniformityAnalysisImpl {
 
   void print(raw_ostream &out) const;
 
+  iterator_range<const UseOutsideCycleInfoT *> uses_outside_cycle() const;
```
Is it correct that this only tracks uses outside a cycle with divergent exits? This should be in the name or at least in a comment.

ssahasra commented on llvm/include/llvm/ADT/GenericUniformityImpl.h:
```diff
@@ -427,6 +432,7 @@ template <typename ContextT> class GenericUniformityAnalysisImpl {
 
   // Recognized cycles with divergent exits.
   SmallPtrSet<const CycleT *, 16> DivergentExitCycles;
+  SmallVector<UseOutsideCycleInfoT, 4> UsesOutsideCycle;
```
Is it correct that this only tracks uses outside a cycle with divergent exits? This should be in the name or at least in a comment.

ruiling reviewed (commented):
I doubt how much help MachineUniformityAnalysis could help here, can it correctly differentiate uniform/divergent booleans stored in SGPR at this late stage? Do you have some tests to show this?

Do you know any other sources of such temporal divergence besides MachineSink?

For the issue shown in the test, I think there might be a little bit more complex case like:
```
bb.loop:
  ...
  %0 = S_ADD ...
  %1 = V_ADD %0, ...
  ...
  SI_LOOP ...
  S_BRANCH %bb.exit
```
Now that we allow `%1` to be sunk, then `%0` which is a scalar instruction may also be sunk out of the loop. So I think inserting a simple sgpr to vgpr copy does not work for such kind of case. If you follow the way shown in the change, you might need to convert the instruction into vector instruction. But I guess there might be constraint that you cannot always do that, which I am not sure about possible issues here.

ruiling commented:
> If all of those can be sunk out of the loop, that would be done in some early ir pass. S_ADD is used by some phi (easiest is to think about it as "loop counter" - the "++i") and cannot be sunk out of the loop

In the example, `%0` is not an induction variable, but a user of induction variable (see below example), it cannot be sunk out of the loop in early ir pass because it has a user `%1` in the loop. Now we allow `%1` to be sunk out of the loop, then `%0` will also be sunk out.
```
bb.loop:
  %ind = PHI %ind_inc, %bb.loop, ...
  %ind_inc = S_ADD %ind, ...
  %0 = S_ADD %ind, ...
  %1 = V_ADD %0, ...
  ...
  SI_LOOP  %bb.loop ...
  S_BRANCH %bb.exit

bb.exit
  ... = %1
```

ssahasra commented:
> I doubt how much help MachineUniformityAnalysis could help here, can it correctly differentiate uniform/divergent booleans stored in SGPR at this late stage? Do you have some tests to show this?

MUA depends on TRI::isUniformReg(), which eventually relies on RBI::isDivergentRegBank(). So yes, MUA can differentiate between a uniform SGPR and a "VCC" type SGPR.

nhaehnle commented:
Can you give a clean explanation of what this new pass does, referring to the definition of MIR semantics and nothing else?

The reason I'm asking is that this entire approach smells like accepting incorrect IR temporarily and then trying to fix it up later. That kind of approach always comes back to bite us in the end. We should not do it.

Relatedly, while trying to understand https://github.com/llvm/llvm-project/blob/3f8ef57bede94445b1a1042c987cc914a886e7ff I've been wondering what the bug in machine-sink-loop-var-out-of-divergent-loop-swdev407790.ll really is/was. The patch changes the location of a definition of v4 from after a loop to inside a loop, but v4 isn't used in the loop as far as I can tell, so why is that change a fix?

petar-avramovic commented:
@nhaehnle  Let's start with https://github.com/llvm/llvm-project/blob/3f8ef57bede94445b1a1042c987cc914a886e7ff
It does not have good regression test (no temporal divergence), and issue it tries to fix was not an issue. However it stopped machine-sinking from sinking outside of the cycle (or after any si_end_cf). 
This way machine-sink can no longer introduce temporal divergent use - "bug fixed", but it can also no longer sink some instructions in cases where it was fine - "new performance bugs".

> Can you give a clean explanation of what this new pass does, referring to the definition of MIR semantics and nothing else?

We need a late pass that fixes sgpr defined inside cycle used outside the cycle with divergent exit.
Fix is to copy sgpr to vgpr inside the cycle and use that vgpr outside the cycle instead.

> The reason I'm asking is that this entire approach smells like accepting incorrect IR temporarily and then trying to fix it up later. That kind of approach always comes back to bite us in the end. We should not do it.

Temporal divergent use is just a simple use outside the cycle in IR, and I think that IR was fine.The temporal divergence errors are visible after register classes are assigned.
MIR is correct after SIFixSGPRCopies.cpp (temporal divergence is fixed by reassigned sgpr Src from inside the cycle to vgpr)

Now if there is a pass that can sink vgpr instruction outside of the cycle and instruction used sgpr Src defined inside the cycle we have a problem. Ideally we should place this pass late in the pipeline when other passes stop moving instructions outside the cycle.

Why this worked before is that everything was sunk outside of the cycles in IR passes and  there was nothing left for machine-sink to sink

nhaehnle commented:
> Now if there is a pass that can sink vgpr instruction outside of the cycle and instruction used sgpr Src defined inside the cycle we have a problem.

Right. It seems like a pass that does this is buggy and *that* should be fixed instead of adding yet another pass that tries to fixup mistakes made by an earlier pass.

petar-avramovic commented:
> Right. It seems like a pass that does this is buggy and that should be fixed instead of adding yet another pass that tries to fixup mistakes made by an earlier pass.

The argument for the separate pass was that generic pass machine-sink would use amdgpu-target-specific hook that says fix-temporal-divergence. Also we would have to re-calculate machine uniformity analysis info after each sink or do it once at the end of the pass (not much different then a separate pass). So I was thinking it was best to catch all temporal divergence cases in late stage of pipeline created by any pass. GlobalISel could use this as its only temporal divergence lowering point.
Code for the target hook would be pretty much the same.

petar-avramovic commented:
I was trying to understand better what was the issue and I think it is not target specific but order of generic IR passes specific.
LICM was moved later in the pipeline https://github.com/llvm/llvm-project/commit/5b657f50b8e8dc5836fb80e566ca7569fd04c26f.
In old days LICM would sink all these add instructions out of the loop but now there are some lcssa phi in the way.
This happens before target specific control flow intrinsics are introduced.
We probably should not depend on optimization passes for correctness.

OK, I will add target hook for this.

petar-avramovic commented:
Target hook version with updated loop->cycle and smaller test https://github.com/llvm/llvm-project/pull/67456.

ruiling commented:
> I understand now, that does look like an issue but I don't have a test for that. I assumed that IR passes would move both %0 and %1 outside the loop.
There are many reasons that a instruction cannot be sunk. For this situation, it is much easier to make a test using mir. I don't think it is too hard to get the MIR output of the test in the PR before machine-sink and add one scalar instruction S_ADD to show the issue.

ruiling commented:
> > I doubt how much help MachineUniformityAnalysis could help here, can it correctly differentiate uniform/divergent booleans stored in SGPR at this late stage? Do you have some tests to show this?
> 
> MUA depends on TRI::isUniformReg(), which eventually relies on RBI::isDivergentRegBank(). So yes, MUA can differentiate between a uniform SGPR and a "VCC" type SGPR.

I think this is for the GlobalISel path? Does the SDAG path also work this way?

petar-avramovic commented:
> > > I doubt how much help MachineUniformityAnalysis could help here, can it correctly differentiate uniform/divergent booleans stored in SGPR at this late stage? Do you have some tests to show this?
> > 
> > 
> > MUA depends on TRI::isUniformReg(), which eventually relies on RBI::isDivergentRegBank(). So yes, MUA can differentiate between a uniform SGPR and a "VCC" type SGPR.
> 
> I think this is for the GlobalISel path? Does the SDAG path also work this way?

If you use sgpr register class + LLT::scalar(1) on all lane mask registers it works. Currently neither path does this. But it is more convenient for GlobalISel since most lane masks already have LLT::scalar(1) before register classes are assigned.

petar-avramovic commented:
> > I understand now, that does look like an issue but I don't have a test for that. I assumed that IR passes would move both %0 and %1 outside the loop.
> 
> There are many reasons that a instruction cannot be sunk. For this situation, it is much easier to make a test using mir. I don't think it is too hard to get the MIR output of the test in the PR before machine-sink and add one scalar instruction S_ADD to show the issue.

Ugh, sgpr reg class is too strong, if you sink sgpr instruction it is still uniform because of 
```
    if (TRI.isUniformReg(MRI, RBI, op.getReg()))
      continue;
```
It was meant to be marked as divergent because of temporal divergence. This breaks uniformity info after the loop.

petar-avramovic commented:
> There are many reasons that a instruction cannot be sunk. For this situation, it is much easier to make a test using mir. I don't think it is too hard to get the MIR output of the test in the PR before machine-sink and add one scalar instruction S_ADD to show the issue.

Added mir test in https://github.com/llvm/llvm-project/pull/67456 
switched to using register classes and opcodes instead of machine-uniformity-analysis-info and left a fixme comment

nhaehnle commented:
So, can this be closed now since it seems to be superseded by #67456?

petar-avramovic commented:
https://github.com/llvm/llvm-project/pull/67456
